File: 2000/2146.cpp
#include<iostream>
#include<queue>
using namespace std;
#define p pair<int, int>
int n, answer = 2e9;
int area[101][101];
int color[101][101];
int dx[]{ 0,1,0,-1 };
int dy[]{ 1,0,-1,0 };

int min(int a, int b) {
	if (a < b) return a;
	return b;
}

int abs(int a) {
	if (a > 0) return a;
	return -a;
}

void input() {
	cin >> n;
	for (int i = 1; i <= n; ++i) {
		for (int j = 1; j <= n; ++j) {
			cin >> area[i][j];
		}
	}
}

void distribute() {
	int cnt = 1;
	for (int i = 1; i <= n; ++i) {
		for (int j = 1; j <= n; ++j) {
			if (area[i][j] == 0 || color[i][j] != 0) continue;
			queue<p> q;
			q.push({ i,j });
			color[i][j] = cnt;
			while (!q.empty()) {
				p cur = q.front(); q.pop();
				for (int dir = 0; dir < 4; ++dir) {
					int x = cur.first + dx[dir];
					int y = cur.second + dy[dir];
					if (x < 1 || y < 1 || x > n || y > n) continue;
					if (color[x][y] != 0 || area[x][y] == 0) continue;
					color[x][y] = cnt;
					q.push({ x,y });
				}
			}
			++cnt;
		}
	}
}

void bfs(int x, int y) {
	queue<p> q;
	q.push({ x,y });
	bool vis[101][101]{ 0 };
	vis[x][y] = 1;
	while (!q.empty()) {
		p cur = q.front(); q.pop();
		for (int dir = 0; dir < 4; ++dir) {
			int nx = cur.first + dx[dir];
			int ny = cur.second + dy[dir];
			if (nx < 1 || ny < 1 || nx > n || ny > n) continue;
			int a = color[nx][ny], b = color[x][y];
			if(a > 0 && b > 0 && a != b){
				answer = min(answer, (abs(nx - x) + abs(ny - y) - 1));
				return;
			}
			if (area[nx][ny] == 1 || vis[nx][ny]) continue;
			q.push({ nx,ny });
			vis[nx][ny] = 1;
		}
	}
}

void solution() {
	input();
	distribute();
	for (int i = 1; i <= n; ++i) {
		for (int j = 1; j <= n; ++j) {
			if (area[i][j] == 0) continue;
			bfs(i, j);
		}
	}
	cout << answer;
}

int main() {
	cin.tie(0), cout.tie(0);
	ios_base::sync_with_stdio(0);
	solution();
	return 0;
}

File: 2000/2146.java
import java.io.*;
import java.util.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringBuilder sb = new StringBuilder();
	  static StringTokenizer st;
    static int n;
    static int[][] area;
    static boolean[][] vis;
    static final int[] dx = {0, 1, 0, -1};
    static final int[] dy = {1, 0, -1, 0};

    static void input() throws IOException {
		st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		area = new int[n][n];
		vis = new boolean[n][n];
    	for(int i = 0;i < n; ++i) {
    		st = new StringTokenizer(br.readLine(), " ");
    		for(int j = 0;j < n; ++j) {
    			area[i][j] = Integer.parseInt(st.nextToken());
    		}
    	}
    }
    
    static boolean range(int x, int y) {
    	return x >= 0 && y >= 0 && x < n && y < n;
    }
    
    static void divideArea() {
    	int cnt = 0;
    	for(int i = 0;i < n; ++i) {
    		for(int j = 0;j < n; ++j) {
    			if(vis[i][j] || area[i][j] == 0) continue;
    			++cnt;
    			Deque<Integer> q = new ArrayDeque<>(); 
    			q.add(i * n + j);
    			vis[i][j] = true;
    			while(!q.isEmpty()) {
    				int cur = q.poll();
    				area[cur / n][cur % n] = cnt;
    				for(int dir = 0;dir < 4; ++dir) {
    					int nx = cur / n + dx[dir];
    					int ny = cur % n + dy[dir];
    					if(!range(nx, ny) || area[nx][ny] == 0) continue;
    					if(vis[nx][ny]) continue;
    					vis[nx][ny] = true;
    					q.add(nx * n + ny);
    				}
    			}
    		}
    	}
    }
    
    static void solution() {
    	divideArea();
    	int answer = (int)1e9;
    	for(int i = 0;i < n; ++i) {
    		for(int j = 0;j < n; ++j) {
    			if(area[i][j] == 0) continue;
    			vis = new boolean[n][n];
    			answer = Math.min(answer, bfs(i, j));
    		}
    	}
    	System.out.println(answer);
    }
    
    static int bfs(int i, int j) {
		Deque<Integer> q = new ArrayDeque<>(); 
		q.add(i * n + j);
		vis[i][j] = true;
		while(!q.isEmpty()) {
			int cur = q.poll();
			for(int dir = 0;dir < 4; ++dir) {
				int nx = cur / n + dx[dir];
				int ny = cur % n + dy[dir];
				if(!range(nx, ny) || vis[nx][ny]) continue;
				if(area[nx][ny] == area[i][j]) continue;
				if(area[nx][ny] > 0 && area[nx][ny] != area[i][j]) {
					return Math.abs(nx - i) + Math.abs(ny - j) - 1;
				}
				vis[nx][ny] = true;
				q.add(nx * n + ny);
			}
		}
    	return (int)1e9;
    }

    public static void main(String[] args) throws IOException {
   		input();
       	solution();
    }
}
